lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
       @namespace = namespace
       @context = context
-      @file = file
       @logger = logger
+      @template = template
```

Comment:
This is more of a representation of the resource in JSON, not a template. WDYT? The name `template` may confuse someone who reads the code.